src/platforms/stlinkv3/usb_dwc_common.c

```diff
@@ -0,0 +1,480 @@
+	}
+}
+
+uint16_t dwc_ep_write_packet(usbd_device *usbd_dev, uint8_t addr, const void *buf, uint16_t len)
+{
+	const uint32_t *buf32 = buf;
+#if defined(__ARM_ARCH_6M__)
+	const uint8_t *buf8 = buf;
+	uint32_t word32;
+#endif /* defined(__ARM_ARCH_6M__) */
+	int i;
+
+	addr &= 0x7f;
+
+	/* Note: because of the libopencm3 API specification of this
+	 * function, the type of the return code is 'uint16_t'.
+	 * This means that it is not possible to return a negative
+	 * result code for error. Also, the API specification
+	 * expects that the length of the data packet written to
+	 * the usb core is returned. This means that zero-length
+	 * usb data packets cannot be reliably sent with the current api,
+	 * and zero-length packets are needed in some cases for usb
+	 * cdcacm class interfaces to signify the end of a transfer.
+	 * At the moment, the blackmagic firmware works around this
+	 * limitation by sending a single byte (containing a zero)
+	 * usb packet for the gdb cdcacm interface, whenever a zero-length
+	 * packet needs to be sent. This particular workaround works
+	 * for the blackmagic use case, because the gdb protocol is
+	 * packetized, and the single zero byte will be silently discarded
+	 * by gdb, but this approach is incorrect in general.
+	 *
+	 * For the time being, return zero here in case of error. */
+	/* Return if endpoint is already enabled, which means a packet
+	 * transfer is still in progress. */
+	if (REBASE(OTG_DIEPCTL(addr)) & OTG_DIEPCTL0_EPENA)
+		return 0;
+	if (REBASE(OTG_DTXFSTS(addr) & 0xffff) < (unsigned)((len + 3) >> 2))
+		return 0;
+
+	/* Enable endpoint for transmission. */
+	REBASE(OTG_DIEPTSIZ(addr)) = OTG_DIEPSIZ0_PKTCNT | len;
+
+	/* WARNING: it is not explicitly stated in the ST documentation,
+	 * but the usb core fifo memory read/write accesses should not
+	 * be interrupted by usb core fifo memory write/read accesses.
+	 *
+	 * For example, this function can be run from within the usb interrupt
+	 * context, and also from outside of the usb interrupt context.
+	 * When this function executes outside of the usb interrupt context,
+	 * if it gets interrupted by the usb interrupt while it writes to
+	 * the usb core fifo memory, and from within the usb
+	 * interrupt the usb core fifo memory is read, then when control returns
+	 * to this function, the usb core fifo memory write accesses can not
+	 * simply continue, this will result in a transaction error on the
+	 * usb line.
+	 *
+	 * In order to avoid such situation, the usb interrupt is masked here
+	 * prior to writing the data to the usb core fifo memory.
+	 */
+	uint32_t saved_interrupt_mask = REBASE(OTG_GINTMSK);
+	REBASE(OTG_GINTMSK) = 0;
+	REBASE(OTG_DIEPCTL(addr)) |= OTG_DIEPCTL0_EPENA | OTG_DIEPCTL0_CNAK;
+
+	/* Copy buffer to endpoint FIFO, note - memcpy does not work.
+	 * ARMv7M supports non-word-aligned accesses, ARMv6M does not. */
+#if defined(__ARM_ARCH_7M__) || defined(__ARM_ARCH_7EM__)
+	for (i = len; i > 0; i -= 4) {
```

Comment:
This and several other of the loops here would be better served and less confusing to be written in the form
```c
for (uint16_t i =0; i < len; i += 4)
```
which would also be less bugprone. let's change them over.

Reply:
It's written this way to support non-word-aligned accesses. We check the remainder of `i` after the for loop to sweep up any leftover bytes

Reply:
Oops, I was looking in the `ep_read_packet_internal()` function. This one seems ok to change